### PR TITLE
Chronam 344, 345, 346

### PR DESCRIPTION
**Pull request by lsukharn**
Please review @myusuf  @johnscancella 
tickets: 
chronam-344(move get_text_from_solr to index.py),
chronam-345(add new unit/integration tests),
chronam-346(drop text field from db) and
chronam-357(modifying batch_loader.py for new db schema)

### CODE REVIEW

File: core/tests/ocr_dump_tests.py [hunk 1]
```diff
@@ -50,7 +51,8 @@ def test_new_dump(self):
         t = tarfile.open(dump.path, "r:bz2")
         members = t.getmembers()
         self.assertEqual(len(members), 112) # ocr xml and txt for each page
-        self.assertEqual(members[0].size, 26282)
+        # Now each record is much shorter because the text field is removed
```
Comment (johnscancella) [inline core/tests/ocr_dump_tests.py]: This comment does not make sense without the context of your tasks. Please remove it.

File: core/batch_loader.py [hunk 2]
```diff
@@ -428,15 +428,17 @@ def process_ocr(self, page, index=True):
...
+        lang_text_solr = {}
         for lang, text in lang_text.iteritems():
             try:
                 language = models.Language.objects.get(Q(code=lang) | Q(lingvoj__iendswith=lang))
             except models.Language.DoesNotExist:
                 # default to english as per requirement
                 language = models.Language.objects.get(code='eng')
-            ocr.language_texts.create(language=language,
-                                      text=text)
+            lang_text_solr[str(language.code)] = text
```
[the remark this answers is not in the document]
Comment (lsukharn) [reply core/batch_loader.py]: Yes, it's a string, I will remove str()

File: core/tests/__init__.py [hunk 3]
```diff
@@ -4,8 +4,9 @@
 from holding_loader_tests import *
 from title_pull_tests import *
 from batch_loader_tests import *
+from browse_tests import *
 from ocr_extractor_tests import *
-from essay_loader_tests import *
+# from essay_loader_tests import *
```
[the remark this answers is not in the document]
Comment (lsukharn) [reply core/tests/__init__.py]: essay_loader_tests don't work now. I assume when essay loader will be fixed these tests will work again. Therefore, I agree, I should remove commenting out from this line

File: core/tests/ocr_dump_tests.py (hunk 1 above)
Comment (lsukharn) [reply core/tests/ocr_dump_tests.py]: I'll remove it

File: core/tests/batch_loader_tests.py [hunk 4]
```diff
@@ -53,7 +53,6 @@ def test_load_batch(self):
         # extract ocr data just for this page
         loader.process_ocr(page, index=False)
         self.assertTrue(page.ocr != None)
-        self.assertTrue(len(page.ocr.text) > 0)
```
[the remark this answers is not in the document]
Comment (lsukharn) [reply core/tests/batch_loader_tests.py]: Good point, I added this line into batch_loader_test.py and retested everything - all tests passed

File: core/index.py [hunk 5]
```diff
@@ -666,3 +666,10 @@ def _solrize_date(d, is_start=True):
     else:
         return None
 
+def get_page_text(page):
+    solr = SolrConnection(settings.SOLR)
+    title, date = page.issue.title, page.issue.date_issued
+    query = 'title:"' + str(title) + '" and date:' + str(date).replace('-', '') + \
+            ' and sequence:' + str(page.sequence)
+    solr_results = solr.query(query)
+    return solr_results.results[0]['ocr']
```
[the remark this answers is not in the document]
Comment (lsukharn) [reply core/index.py]: I agree, thanks. I am going to add code that checks whether solr_results.results is empty

File: core/models.py [hunk 6]
```diff
@@ -1207,12 +1204,14 @@ def __unicode__(self):
...
     def _add_page(self, page, tar):
+        from index import get_page_text
+
         d = page.issue.date_issued
         relative_dir = "%s/%i/%02i/%02i/ed-%i/seq-%i/" % (page.issue.title_id, d.year, d.month, d.day, page.issue.edition, page.sequence)
 
         # add ocr text
         txt_filename = relative_dir + "ocr.txt"
-        ocr_text = page.ocr.text.encode('utf-8')
+        ocr_text = get_page_text(page)[0].encode('utf-8')
```
Comment (johnscancella) [inline core/models.py]: How do you know that `get_page_text(page)` will be populated and not throw an array index exception?

File: core/tests/browse_tests.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+from django.db import connection
+from django.conf import settings
+
+from chronam.core.views.browse import get_page_text
+from chronam.core.utils.utils import _get_tip
+from chronam.core.batch_loader import BatchLoader, Batch
+from chronam.core.models import Title, LanguageText
+
+
+class BrowseTests(TestCase):
```
Comment (johnscancella) [inline core/tests/browse_tests.py]: Why is this called BrowseTests? Perhaps put a class comment describing what this tests?

File: core/models.py (hunk 6 above)
Comment (lsukharn) [reply core/models.py]: New implementation of get_page_text precludes returning of an empty array. It now returns ['Text not available'] even if there is no ocr for the page. If you see that it's not the case please let me know where you think the function can return an empty array or None. In the meantime I am going to add a length check condition (in core/models.py line 1214) on results of get_page_text(page) to make it even safer

File: core/tests/browse_tests.py (hunk 7 above)
Comment (lsukharn) [reply core/tests/browse_tests.py]: Yes, I will add a class comment there

File: core/models.py [hunk 8]
```diff
@@ -1207,12 +1204,18 @@ def __unicode__(self):
         return "path=%s size=%s sha1=%s" % (self.path, self.size, self.sha1)
 
     def _add_page(self, page, tar):
+        from index import get_page_text
```
Comment (johnscancella) [inline core/models.py]: This belongs at the top of the file

File: core/tests/browse_tests.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Tests related to core/views/browse.py.
+    test_full_text_deleted - 'sanity' check that 'text' column is removed
```
Comment (johnscancella) [inline core/tests/browse_tests.py]: Could you move line 16, 17, and 18 to their respective tests, and be more descriptive then `test_getting_text_from_solr_utah and test_getting_text_from_solr_slovenia`?

File: core/models.py (hunk 8 above)
Comment (lsukharn) [reply core/models.py]: No, it doesn't. If it will be at the top of the file a circular import will cause and error

File: core/tests/browse_tests.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_full_text_deleted(self):
+        #'sanity' check that 'text' column is removed
+        check = False
+        with connection.cursor() as cursor:
+            cursor.execute("SHOW COLUMNS FROM core_languagetext")
+            rows = cursor.fetchall()
+            for row in rows:
+                if row[0] == 'text':
+                    check = True
```
Comment (johnscancella) [inline core/tests/browse_tests.py]: does it make more sense to use `self.fail("core_languagetext.text should not exist. Did you run the migrations?")` here? That would allow you to remove line 21 and 28

File: core/models.py (hunk 8 above)
Comment (acdha) [reply core/models.py]: One minor note: it'd be better style to use an explicitly relative import in case anything in the search path ever adds a module named `index`: `from .index import get_page_text`.

On a larger note, that would definitely cause a circular import and that suggests that this code might belong somewhere else, especially since this is doing more file I/O than database work and this method and the method which calls it are invoked by exactly one point (the `dump_ocr` task). It seems like it'd be better just to move that code into the task and keep the models clean, which would also keep the import flow consistent.

File: core/models.py [hunk 11]
```diff
@@ -1207,12 +1204,18 @@ def __unicode__(self):
...
         relative_dir = "%s/%i/%02i/%02i/ed-%i/seq-%i/" % (page.issue.title_id, d.year, d.month, d.day, page.issue.edition, page.sequence)
 
         # add ocr text
         txt_filename = relative_dir + "ocr.txt"
-        ocr_text = page.ocr.text.encode('utf-8')
+        ocr_text = get_page_text(page)
+        if len(ocr_text) > 0:
+            ocr_text = ocr_text[0].encode('utf-8')
+        else:
+            ocr_text = ['Text not available']
```
Comment (acdha) [inline core/models.py]: This duplicates the constant returned by `get_page_text`. It isn't immediately obvious that this code path can even be called because if I'm reading https://github.com/lsukharn/chronam/blob/267daf1fb713353d70fd7a52043bd4ce407a2515/core/index.py#L669-L680 correctly `get_page_text` will always return a non-empty list.

File: core/models.py [hunk 12]
```diff
@@ -1207,12 +1204,18 @@ def __unicode__(self):
...
+        from index import get_page_text
+
         d = page.issue.date_issued
         relative_dir = "%s/%i/%02i/%02i/ed-%i/seq-%i/" % (page.issue.title_id, d.year, d.month, d.day, page.issue.edition, page.sequence)
 
         # add ocr text
         txt_filename = relative_dir + "ocr.txt"
-        ocr_text = page.ocr.text.encode('utf-8')
+        ocr_text = get_page_text(page)
+        if len(ocr_text) > 0:
```
Comment (acdha) [inline core/models.py]: This can simply be `if ocr_text:`

File: core/models.py (hunk 11 above)
Comment (lsukharn) [reply core/models.py]: That's what I commented before, so should I keep this code or return to the previous implementation? (I wrote get_page_text with an idea that it will always returns non-empty list)

File: core/tests/batch_loader_tests.py [hunk 13]
```diff
@@ -53,7 +53,7 @@ def test_load_batch(self):
...
+        self.assertTrue(len(page.lang_text) > 0)
```
Comment (acdha) [inline core/tests/batch_loader_tests.py]: Use [`assertGreater`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertGreater) so if the test fails it will print the values rather than just true/false.

If this is coming from a test fixture, you could even use [`assertListEqual`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertListEqual) to confirm the expected value.

File: core/tests/batch_loader_tests.py [hunk 14]
```diff
@@ -85,8 +85,10 @@ def test_load_batch(self):
             'uuml_thys_ocr.txt')
         self.assertEqual(solr_doc['ocr_eng'], file(f).read().decode('utf-8'))
 
-
         # purge the batch and make sure it's gone from the db
         loader.purge_batch('batch_uuml_thys_ver01')
         self.assertEqual(Batch.objects.all().count(), 0)
         self.assertEqual(Title.objects.get(lccn='sn83045396').has_issues, False)
+
+
```
Comment (acdha) [inline core/tests/batch_loader_tests.py]: We should remove the extra lines here following PEP-8

File: core/models.py (hunk 11 above)
Comment (johnscancella) [reply core/models.py]: yes, revert to what you had before. I wasn't following as closely as I should have, and as it was pointed it out, it will never be None

File: core/tests/browse_tests.py (hunk 10 above)
Comment (acdha) [reply core/tests/browse_tests.py]: In a test, yes

**Comment by acdha** [comment]
I would recommend rebasing your working branch against the current master branch unless multiple other people are actively making commits in your working branch. Rebasing avoids the need to have merge commits & will allow you to update your existing work against the new upstream so you won't need separate commits to reconcile changes.

File: core/tests/batch_loader_tests.py (hunk 13 above)
Comment (lsukharn) [reply core/tests/batch_loader_tests.py]: I added assertGreater here. Thanks

File: core/tests/batch_loader_tests.py (hunk 14 above)
Comment (lsukharn) [reply core/tests/batch_loader_tests.py]: Removed the extra lines

File: core/models.py (hunk 12 above)
Comment (lsukharn) [reply core/models.py]: Now reverted to ocr_text = get_page_text(page)[0].encode('utf-8')

File: core/models.py (hunk 11 above)
Comment (lsukharn) [reply core/models.py]: Reverted to ocr_text = get_page_text(page)[0].encode('utf-8')

File: core/tests/browse_tests.py (hunk 10 above)
Comment (lsukharn) [reply core/tests/browse_tests.py]: Changed browse_tests according to the suggestion

File: core/models.py (hunk 8 above)
Comment (lsukharn) [reply core/models.py]: I agree, I am going to create a new ticket for it

File: core/index.py [hunk 15]
```diff
@@ -666,3 +666,15 @@ def _solrize_date(d, is_start=True):
     else:
         return None
 
+def get_page_text(page):
+   no_text = ["Text not available"]
+   solr = SolrConnection(settings.SOLR)
+   title, date = page.issue.title, page.issue.date_issued
+   query = 'title:"' + str(title) + '" and date:' + str(date).replace('-', '') + \
```
Comment (acdha) [inline core/index.py]: This should use string formatting. Also, it looks like the inputs need to be escaped — e.g. that query looks like it would break if `title` had a `"` character or anything else with meaning to Solr.

File: core/index.py [hunk 16]
```diff
@@ -666,3 +666,15 @@ def _solrize_date(d, is_start=True):
...
+           ' and sequence:' + str(page.sequence)
+   solr_results = solr.query(query)
+   results_attribute = getattr(solr_results, 'results', None)
+   if results_attribute and len(results_attribute) > 0:
```
Comment (acdha) [inline core/index.py]: This test is redundant as written – both `None` and `[]` will fail the `if results_attribute` check. If the idea is to test whether it's a value other than either `None` or a `list`, I would do an `isinstance` check instead.

File: core/tests/ocr_dump_tests.py [hunk 17]
```diff
@@ -34,8 +34,9 @@ def test_new_dump(self):
         self.assertEqual(dump.path, os.path.join(dumps_dir, "batch_uuml_thys_ver01.tar.bz2"))
         # size can actually vary based on the compression of the different dates
         # that are in the tarfile
-        self.assertTrue(dump.size > 3500000)
-        self.assertTrue(dump.size < 4500000)
+        self.assertTrue(dump.size > 3000000)
```
Comment (acdha) [inline core/tests/ocr_dump_tests.py]: `assertIsGreater` here too

File: core/views/browse.py [hunk 18]
```diff
@@ -23,6 +23,7 @@
 from chronam.core.utils.url import unpack_url_path
 from chronam.core import models, index
 from chronam.core.rdf import title_to_graph, issue_to_graph, page_to_graph
+from chronam.core.index import get_page_text
```
Comment (acdha) [inline core/views/browse.py]: Nothing wrong with this import but it would be nice to have them sorted using something like `isort` for readability.

File: core/views/browse.py [hunk 19]
```diff
@@ -504,7 +506,7 @@ def page_ocr_xml(request, lccn, date, edition, sequence):
 def page_ocr_txt(request, lccn, date, edition, sequence):
     title, issue, page = _get_tip(lccn, date, edition, sequence)
     try:
-        text = page.ocr.text
+        text = get_page_text(page)
         return HttpResponse(text, content_type='text/plain')
```
Comment (acdha) [inline core/views/browse.py]: As a style point, I'd prefer to move the return below the try block since that exception won't be thrown by the second statement.

File: loc/templates/page.html [hunk 20]
```diff
@@ -19,7 +19,9 @@
 
     <noscript><!-- without javascript the pageviewer does not work; display OCR in this case instead. -->
         <h3>Newspaper Page Text</h3>
-        <pre>{{ page.ocr.text }}</pre>
+        {% if text %}
+        <pre>{{ text.0|linebreaks }} </pre>
```
Comment (acdha) [inline loc/templates/page.html]: Is this really just text.0 or is it possible to have more than one line, where we'd need something like `|join`?

File: loc/templates/page_text.html [hunk 21]
```diff
@@ -38,11 +38,13 @@
     </div>
 
     <div>
-    {{page.ocr.text|escape|linebreaks}}
+    {% if text %}
+    {{ text.0|linebreaks }}
```
Comment (acdha) [inline loc/templates/page_text.html]: Same question about whether this can possibly have more than one element

File: core/views/browse.py (hunk 18 above)
Comment (lsukharn) [reply core/views/browse.py]: I agree, but can it be done later, maybe as a part of a different ticket?

File: core/index.py (hunk 15 above)
Comment (lsukharn) [reply core/index.py]: Added string formatting and escaping for quotes in title

File: core/index.py (hunk 16 above)
Comment (lsukharn) [reply core/index.py]: added isinstance(results_attribute, list)

File: core/tests/ocr_dump_tests.py (hunk 17 above)
Comment (lsukharn) [reply core/tests/ocr_dump_tests.py]: changed to assertGreater

File: loc/templates/page.html (hunk 20 above)
Comment (lsukharn) [reply loc/templates/page.html]: added join filter to the template

File: loc/templates/page_text.html (hunk 21 above)
Comment (lsukharn) [reply loc/templates/page_text.html]: added join filter to the template

File: core/views/browse.py (hunk 19 above)
Comment (lsukharn) [reply core/views/browse.py]: moved below

**Comment by lsukharn** [comment]
If there is no more code change suggestions please merge so I can close the tickets related to this pull request

File: core/views/browse.py (hunk 18 above)
Comment (acdha) [reply core/views/browse.py]: Why not add a commit before merging this branch? It should be quick.

File: core/index.py [hunk 22]
```diff
@@ -670,11 +670,12 @@ def get_page_text(page):
    no_text = ["Text not available"]
    solr = SolrConnection(settings.SOLR)
    title, date = page.issue.title, page.issue.date_issued
-   query = 'title:"' + str(title) + '" and date:' + str(date).replace('-', '') + \
-           ' and sequence:' + str(page.sequence)
+   query = 'title: %s and date:%s and sequence:%s' % (str(title).replace('"', ''),
```
Comment (acdha) [inline core/index.py]: Does solrpy not have a function to escape query components? Based on [https://lucene.apache.org/core/2_9_4/queryparsersyntax.html#Escaping%20Special%20Characters](the Lucene documentation) there are other characters which we should escape as well. If Solrpy doesn't have a function which does that, we could add a little utility.

File: core/views/browse.py [hunk 23]
```diff
@@ -507,9 +507,10 @@ def page_ocr_txt(request, lccn, date, edition, sequence):
     title, issue, page = _get_tip(lccn, date, edition, sequence)
     try:
         text = get_page_text(page)
-        return HttpResponse(text, content_type='text/plain')
     except models.OCR.DoesNotExist:
-        raise Http404("No OCR for %s" % page)
+        raise Http404("No OC10dhyR for %s" % page)
```
Comment (acdha) [inline core/views/browse.py]: Was this change intentional?

File: loc/templates/page.html [hunk 24]
```diff
@@ -20,7 +20,7 @@
     <noscript><!-- without javascript the pageviewer does not work; display OCR in this case instead. -->
         <h3>Newspaper Page Text</h3>
         {% if text %}
-        <pre>{{ text.0|linebreaks }} </pre>
+        <pre>{{ text|join:" // " |linebreaks }} </pre>
```
Comment (acdha) [inline loc/templates/page.html]: Should this be `//` or `\n`?

File: loc/templates/page_text.html [hunk 25]
```diff
@@ -39,7 +39,7 @@
 
     <div>
     {% if text %}
-    {{ text.0|linebreaks }}
+    {{ text|join:" // "|linebreaks }}
```
Comment (acdha) [inline loc/templates/page_text.html]: Same question re:`//` or `\n`

File: core/index.py (hunk 22 above)
Comment (lsukharn) [reply core/index.py]: My bad, there is actually solr_escape that I just used

File: core/views/browse.py (hunk 23 above)
Comment (lsukharn) [reply core/views/browse.py]: oh, thanks for catching it!

File: core/views/browse.py (hunk 18 above)
Comment (lsukharn) [reply core/views/browse.py]: I applied isort

Comment (acdha) [reply core/views/browse.py]: Thanks!

File: core/index.py (hunk 22 above)
Comment (acdha) [reply core/index.py]: Thank you

File: loc/templates/page_text.html (hunk 25 above)
Comment (lsukharn) [reply loc/templates/page_text.html]: I added <br><br>. Please see recent commit

File: loc/templates/page.html (hunk 24 above)
Comment (lsukharn) [reply loc/templates/page.html]: I added \n\n. Please see recent commit

File: core/models.py (hunk 8 above)
Comment (johnscancella) [reply core/models.py]: Why do you need a new ticket? We are expressing concern here, you should leave the code cleaner than it was.

Comment (acdha) [reply core/models.py]: When I was reviewing this with Leo, he suggested a separate issue and I agreed because the change is only tangentially related to this work. Since it's small and isolated that could be an argument for a single commit here but I was hoping to keep more of a searchable history (the title on this PR bothers me for the same reasons) with cross-references between the issues.

Comment (johnscancella) [reply core/models.py]: I suppose, I guess I see it more as cleanup than a separate task. I have always taken the liberty of cleaning up code while I am poking around other things. This stems from the fact that when I worked at Boeing we weren't allowed to have tasks just to clean up code, so it was understood you had to practice "leave your code cleaner than you left it", otherwise things would just become unmaintainable.

Comment (acdha) [reply core/models.py]: To be clear, I don't see this as an especially strong issue — it _definitely_ should be a separate commit but since we're talking about two methods called in one place & couple dozen lines of code there's a pretty low upper bar for how strongly I'd argue for either option.
